druid-shell/src/gtk/mod.rs

```diff
@@ -0,0 +1,689 @@
+                    );
+                }
+
+                Inhibit(true)
+            });
+        }
+
+        {
+            let handle = handle.clone();
+            drawing_area.connect_motion_notify_event(move |_widget, motion| {
+                if let Some(state) = handle.state.upgrade() {
+                    let mut ctx = WinCtxImpl::from(&handle);
+
+                    let pos = Point::from(motion.get_position());
+                    let mouse_event = window::MouseEvent {
+                        pos,
+                        mods: get_modifiers(motion.get_state()),
+                        count: 0,
+                        button: get_mouse_button_from_modifiers(motion.get_state()),
+                    };
+
+                    state
+                        .handler
+                        .borrow_mut()
+                        .mouse_move(&mouse_event, &mut ctx);
+                }
+
+                Inhibit(true)
+            });
+        }
+
+        {
+            let handle = handle.clone();
+            drawing_area.connect_scroll_event(move |_widget, scroll| {
+                if let Some(state) = handle.state.upgrade() {
+                    let mut ctx = WinCtxImpl::from(&handle);
+
+                    let _deltas = scroll.get_scroll_deltas();
+                    // TODO use these deltas (for smooth scrolling)
+                    let modifiers = get_modifiers(scroll.get_state());
+
+                    // The magic "120"s are from Microsoft's documentation for WM_MOUSEWHEEL.
+                    // They claim that one "tick" on a scroll wheel should be 120 units.
+                    // GTK simply reports the direction
+                    let mut handler = state.handler.borrow_mut();
+                    match scroll.get_direction() {
+                        ScrollDirection::Up => {
+                            handler.wheel(Vec2::from((0.0, -120.0)), modifiers, &mut ctx);
+                        }
+                        ScrollDirection::Down => {
+                            handler.wheel(Vec2::from((0.0, 120.0)), modifiers, &mut ctx);
+                        }
+                        ScrollDirection::Left => {
+                            // Note: this direction is just a guess, I (bobtwinkles) don't
+                            // have a way to test horizontal scroll events under GTK.
+                            // If it's wrong, the right direction also needs to be changed
+                            handler.wheel(Vec2::from((120.0, 0.0)), modifiers, &mut ctx);
+                        }
+                        ScrollDirection::Right => {
+                            handler.wheel(-Vec2::from((-120.0, 0.0)), modifiers, &mut ctx);
+                        }
+                        ScrollDirection::Smooth => {
+                            eprintln!(
+                                "Warning: somehow the Druid widget got a smooth scroll event"
+                            );
+                        }
```

Comment:
This will always be triggered with a touchpad I think - at least my laptop's one emits `ScrollDirection::Smooth`.